### DISCUSSION TITLE
botan binding to Perl/SPVM

### DISCUSSION THREAD

**Discussion by yuki-kimoto** [randombit/botan#2995] [General]
Hi, I'm searching C or C++ library for MD5, SHA, Base64 to bind it to Perl/SPVM.

I found botan today.

Cound I ask some questions?

**Comment by randombit** [comment]
Sure. For what it is worth there was a [Perl binding](https://github.com/randombit/botan/tree/1e93b85fb5a4d628b80a9812b38b842b5c8b4363/src/contrib/perl-xs) but it was removed several years ago as it was unmaintained. In the future anything like that would probably be better done as a separate repo maintained by someone who cares about it.

Any new binding would be better off using the C API (unlike the old Perl binding which used the C++ API directly) since that is guaranteed to be long term stable.

**Comment by yuki-kimoto** [comment]
Thank you for reply.

>  In the future anything like that would probably be better done as a separate repo maintained by someone who cares about it.

OK.

I want to create a separate repo such as <code>SPVM::Resource::Zlib::V1_2_11</code>

https://metacpan.org/pod/SPVM::Resource::Zlib::V1_2_11

> Any new binding would be better off using the C API

OK.

Could I ask some questions about <code>botan</code> to create a binding module?

**Reply by randombit** [reply]
Of course. You may find the [existing bindings](https://github.com/randombit/botan/wiki/Language-Bindings) a useful guide as well.

**Comment by yuki-kimoto** [comment]
Thanks!

I write some questions.

1. What is most stable version?

I want to know this because I create a binding of <b>a</b> version of botan.

2. Does the most stable version of botan need which gcc version?

I want to know this because botan works well on  as many environments as possible.

3. Where are all source codes and headers of C and C++.

I want to know this because I need to write the configure, compilation and link process by myself.

4. Are there the document of the compilation and link options such as C<-DFoo>?

I want to know this because I need to write the configure, compilation and link process by myself.

**Reply by reneme** [reply]
> What is most stable version?

Botan tags releases on a regular basis, and these can be considered stable as long as they are not suffixed with `-alpha`, `-beta` or the like. Currently, that would be 2.19.2. Please find all the tags here: https://github.com/randombit/botan/tags

> Does the most stable version of botan need which gcc version?

The 2.x branch of Botan (which is currently considered stable) relies on C++11 and therefore does not need a very recent compiler. GCC 4.8 should be good enough. More information is available here: https://botan.randombit.net/handbook/support.html

> Where are all source codes and headers of C and C++

You'll find all the source code in Botan's GitHub repository in the `src/lib` directory. Lot's of information how the code is structured is available here: https://botan.randombit.net/handbook/dev_ref/contributing.html
I would highly encourage you to use Botan's build system to build the library, though. Please find information about the inner workings of Botan's build system here: https://botan.randombit.net/handbook/dev_ref/configure.html

> Are there the document of the compilation and link options such as C<-DFoo>?

Yes! Here: https://botan.randombit.net/handbook/building.html

**Reply by yuki-kimoto** [reply]
Thank you very much.

> Are there the document of the compilation and link options such as C<-DFoo>?
> Yes! Here: https://botan.randombit.net/handbook/building.html

Is python is required to compile botan?

**Reply by reneme** [reply]
Botan maintains a custom build system that is written in Python. So, strictly, for compilation python is not required, but for configuring the build (i.e. preparing the `Makefile`) Botan uses Python.

**Reply by yuki-kimoto** [reply]
Thanks.

Without ./configure.py, a library of Botan can be compiled?

For example

```
g++ -c -o hash/md5/md5.o -fPIC hash/md5/md5.cpp 
g++ -shared -o hash/md5/md5.so hash/md5/md5.o
```

**Reply by reneme** [reply]
> Without ./configure.py, a library of Botan can be compiled?

Technically, yes. But it's complicated to get the dependencies of all the compilation units right.

I'm assuming you are targetting some restricted environment where you must rely on doing the compilation yourself?

You could try running the configuration script in a "normal" environment, using `./configure.py`. This results in a `Makefile` containing a list of the compiler/linker calls that need to be performed to obtain a working library. And then try to customize this `Makefile` to your needs and run the commands baremetal on your target environment.

This is nothing Botan would support out-of-the-box but there's a good chance of success.

As a piece of advice: If you go down that path, try configuring with `./configure.py --minimized-build`. This will strip the library's module selection down to the smallest usable set. It'll make the `Makefile` much smaller and easier to digest. From there, you can add additional modules with `--enable-modules=`, e.g.:

```
python3 ./configure.py --minimized-build --enable-modules='md5,sha1'
```

**Reply by yuki-kimoto** [reply]
> I'm assuming you are targetting some restricted environment where you must rely on doing the compilation yourself?

Yes.

I'm trying to create binding module of C/C++ libraries on Perl/SPVM.

**make and Makefile are not used**. Only gcc/g++(MinGW on Windows) is used. 

The information of configuration is got from Perl for example $Config{cc} or $Config{ld}, etc.

Created Examples are:

zlib (C)

https://github.com/yuki-kimoto/SPVM-Resource-Zlib-V1_2_11

Google RE2 (C++)

https://github.com/yuki-kimoto/SPVM-Resource-Re2-V2022_06_01

https://metacpan.org/pod/SPVM::Regex

These product provide Makefile, however manually compliation and link is easy.

Can Botan do this?

**Reply by reneme** [reply]
Botan can be configured into an "amalgamation". That basically combines all the source files into one big `botan_all.cpp` along with a single `botan_all.h` header. Maybe that might make things a lot easier for you:

1. Run `python3 ./configure.py --amalgamation --disable-shared --disable-modules=pkcs11` once
   (This will generate the `botan_all.cpp/h` files in the root of the source tree)
2. Run `g++ -std=c++17 -O3 -DBOTAN_IS_BEING_BUILT -c botan_all.cpp -o botan_all.o`
   (This might run a few seconds or even minutes)
3. Run `ar crs libbotan-3.a botan_all.o` to package the compilation result into a static library

Note that the first step is using Python, but it can be run on any machine. Afterwards you can just copy the `botan_all.cpp/h` files and use plain build tools.

**Reply by yuki-kimoto** [reply]
Thank you very much! I will try Botan binding to Perl/SPVM using <code>botan_all.cpp</code>

**Reply by yuki-kimoto** [reply]
Sorry. I have a question.

Where are header files for <code>botan_all.cpp</code>?

**Reply by reneme** [reply]
Once you run `./configure.py --amalgamation ...` the header should also be amalgamated and placed right next to the cpp file as `botan_all.h`.

**Reply by yuki-kimoto** [reply]
OK.
